velox/common/memory/SharedArbitrator.cpp

```diff
@@ -169,67 +146,157 @@ const SharedArbitrator::Candidate& findCandidateWithLargestCapacity(
       succinctBytes(reclaimableBytes),
       succinctBytes(freeBytes));
 }
 
 SharedArbitrator::~SharedArbitrator() {
-  if (freeCapacity_ != capacity_) {
+  if (freeReservedCapacity_ != reservedCapacity_) {
+    const std::string errMsg = fmt::format(
+        "There is unexpected free reserved capacity not given back to arbitrator "
+        "on destruction: freeReservedCapacity_{} != reservedCapacity_{}\\n{}",
+        freeReservedCapacity_,
+        reservedCapacity_,
+        toString());
+    if (checkUsageLeak_) {
+      VELOX_FAIL(errMsg);
+    } else {
+      VELOX_MEM_LOG(ERROR) << errMsg;
+    }
+  }
+  if (freeNonReservedCapacity_ + freeReservedCapacity_ != capacity_) {
     const std::string errMsg = fmt::format(
-        "\"There is unexpected free capacity not given back to arbitrator "
-        "on destruction: freeCapacity_ != capacity_ ({} vs {})\\n{}\"",
-        freeCapacity_,
+        "There is unexpected free capacity not given back to arbitrator "
+        "on destruction: freeNonReservedCapacity_[{}] + freeReservedCapacity_[{}] != capacity_[{}])\\n{}",
+        freeNonReservedCapacity_,
+        freeReservedCapacity_,
         capacity_,
         toString());
     if (checkUsageLeak_) {
       VELOX_FAIL(errMsg);
     } else {
-      LOG(ERROR) << errMsg;
+      VELOX_MEM_LOG(ERROR) << errMsg;
     }
   }
 }
 
+std::vector<SharedArbitrator::Candidate> SharedArbitrator::getCandidateStats(
+    const std::vector<std::shared_ptr<MemoryPool>>& pools,
+    bool freeCapacityOnly) {
+  std::vector<SharedArbitrator::Candidate> candidates;
+  candidates.reserve(pools.size());
+  for (const auto& pool : pools) {
+    candidates.push_back(
+        {freeCapacityOnly ? 0 : reclaimableUsedCapacity(*pool),
+         reclaimableFreeCapacity(*pool),
+         pool->currentBytes(),
+         pool.get()});
+  }
+  return candidates;
+}
+
+void SharedArbitrator::updateFreeCapacityMetrics() const {
+  RECORD_METRIC_VALUE(
+      kMetricArbitratorFreeCapacityBytes,
+      freeNonReservedCapacity_ + freeReservedCapacity_);
+  RECORD_METRIC_VALUE(
+      kMetricArbitratorFreeReservedCapacityBytes, freeReservedCapacity_);
+}
+
+int64_t SharedArbitrator::reclaimableCapacity(const MemoryPool& pool) const {
+  return std::max<int64_t>(0, pool.capacity() - memoryPoolReservedCapacity_);
+}
+
+int64_t SharedArbitrator::reclaimableFreeCapacity(
+    const MemoryPool& pool) const {
+  return std::min<int64_t>(pool.freeBytes(), reclaimableCapacity(pool));
+}
+
+int64_t SharedArbitrator::reclaimableUsedCapacity(
+    const MemoryPool& pool) const {
+  const auto maxReclaimableBytes = reclaimableCapacity(pool);
+  const auto reclaimableBytes = pool.reclaimableBytes();
+  return std::min<int64_t>(maxReclaimableBytes, reclaimableBytes.value_or(0));
+}
+
+int64_t SharedArbitrator::minGrowCapacity(const MemoryPool& pool) const {
+  return std::max<int64_t>(
+      0,
+      std::min<int64_t>(pool.maxCapacity(), memoryPoolReservedCapacity_) -
```

Comment:
Curious that we do not want to enforce memory pool's max capacity to be larger than reserved capacity?